File: src/main/scala/category/backtrack/_306_Solution.scala
package category.backtrack

/**
  * 306. Additive Number
  *
  * "112358"
  * "199100199"
  * "9999110000100001"
  * "9999110000100001"
  * "9999110000100001100001100001100001"
  * "99999999999999999999999"
  * "1023"
  * "1203"
  * "0123"
  * "123"
  * "199001200"
  * "101"
  * "011"
  * "0011"
  */
object _306_Solution {
  def isAdditiveNumber(num: String): Boolean = {
    
    def concate(s1: String, to: Int): String = {
      val sb = new StringBuilder
      for (i <- 0 until to) sb.append('0')
      sb.append(s1).toString()
    }
    
    def add(s1: String, s2: String): String = {
      val sb = new StringBuilder
      val (update, other) = if (s1.length < s2.length) {
        concate(s1, s2.length - s1.length) -> s2
      } else {
        concate(s2, s1.length - s2.length) -> s1
      }
      var carry = 0
      for (i <- Range(update.length - 1, -1, -1)) {
        val sum = update(i) - '0' + other(i) - '0' + carry
        sb.append(sum % 10)
        carry = sum / 10
      }
      if (carry != 0) sb.append(carry)
      sb.reverse.toString()
    }
    
    def backtrack(input: String, first: String, second: String): Boolean = {
      if (input.isEmpty) return true
      if (first.nonEmpty && second.nonEmpty) {
        val target = add(first, second)
        if (input.indexOf(target) != 0) return false
        backtrack(input.substring(target.length), second, target)
      } else if (first.isEmpty) {
        for (i <- 1 until input.length / 2 + 1) {
          val ans = backtrack(input.substring(i), input.substring(0, i), "")
          if (ans) return true
        }
        false
      } else {
        val max = input.length - first.length
        for (i <- 1 until Math.min(max, input.length / 2) + 1) {
          val second = input.substring(0, i)
          if (second=="0" || !second.startsWith("0")) {
            val ans = backtrack(input.substring(i), first, input.substring(0, i))
            if (ans) return true
          }
        }
        false
      }
    }
    
    backtrack(num, "", "")
  }
}

File: src/main/python/medium/_300_400/_306_Solution.py
# 306. Additive Number
class _306_Solution:
    def isAdditiveNumber(self, num: str) -> bool:
        def addStr(l: str, r: str) -> str:
            n0, n1 = len(l), len(r)
            size = max(n0, n1)
            s0, s1 = l.zfill(size), r.zfill(size)
            carry = 0
            ans = ""
            for i in range(size - 1, -1, -1):
                carry, remain = divmod(int(s0[i]) + int(s1[i]) + carry, 10)
                ans += str(remain)
            if carry != 0: ans += str(carry)
            return ans[::-1]

        def backtrack(left: str, right: str, index: int) -> bool:
            if index >= len(num):
                return True
            sumOfStr = addStr(left, right)
            if sumOfStr == num[index:index + len(sumOfStr)]:
                return backtrack(right, sumOfStr, index + len(sumOfStr))
            else:
                return False

        for i in range(1, len(num) // 2 + 1):
            if num[0:i].startswith("0") and num[0:i] != "0": continue
            for j in range(i + 1, len(num)):
                if num[i:j].startswith("0") and num[i:j] != "0": continue
                if backtrack(num[0:i], num[i:j], j): return True
        return False


'''
"112358"
"199100199"
"214748367812147483679"
"011"
"10111"
"10235"
"1203"
"199001200"
"1023"
"011"
"001199200"
"100111"
'''

if __name__ == '__main__':
    instance = _306_Solution()
    assert not instance.isAdditiveNumber("100111")
    assert not instance.isAdditiveNumber("1023")
    assert instance.isAdditiveNumber("011")
    assert not instance.isAdditiveNumber("001199200")
    assert not instance.isAdditiveNumber("199001200")
    assert instance.isAdditiveNumber("214748367812147483679")
    assert instance.isAdditiveNumber("011")
    assert instance.isAdditiveNumber("10111")
    assert not instance.isAdditiveNumber("1203")
    assert not instance.isAdditiveNumber("10235")
    assert instance.isAdditiveNumber("112358")
    assert instance.isAdditiveNumber("199100199")
